internal/upstreamldap/upstreamldap.go

```diff
@@ -0,0 +1,386 @@
+// Copyright 2021 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package upstreamldap implements an abstraction of upstream LDAP IDP interactions.
+package upstreamldap
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/go-ldap/ldap/v3"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/user"
+
+	"go.pinniped.dev/internal/plog"
+)
+
+const (
+	ldapsScheme                                 = "ldaps"
+	distinguishedNameAttributeName              = "dn"
+	userSearchFilterInterpolationLocationMarker = "{}"
+)
+
+// Conn abstracts the upstream LDAP communication protocol (mostly for testing).
+type Conn interface {
+	Bind(username, password string) error
+
+	Search(searchRequest *ldap.SearchRequest) (*ldap.SearchResult, error)
+
+	Close()
+}
+
+// Our Conn type is subset of the ldap.Client interface, which is implemented by ldap.Conn.
+var _ Conn = &ldap.Conn{}
+
+// LDAPDialer is a factory of Conn, and the resulting Conn can then be used to interact with an upstream LDAP IDP.
+type LDAPDialer interface {
+	Dial(ctx context.Context, hostAndPort string) (Conn, error)
+}
+
+// LDAPDialerFunc makes it easy to use a func as an LDAPDialer.
+type LDAPDialerFunc func(ctx context.Context, hostAndPort string) (Conn, error)
+
+var _ LDAPDialer = LDAPDialerFunc(func(ctx context.Context, hostAndPort string) (Conn, error) { return nil, nil })
```

Comment:
I _think_ this can also just be like:
```suggestion
var _ LDAPDialer = LDAPDialerFunc(nil)
```

Reply:
Fixed in 5c62a9d